# Werkcollege-opdrachten Week 1.3

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige werkcollege de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [1]:
import pandas as pd
import sqlite3
import warnings
warnings.simplefilter('ignore')

Zet de volgende bestanden in een makkelijk te vinden map, als je dat nog niet gedaan hebt:
- go_sales.sqlite
- go_crm.sqlite
- go_staff.sqlite

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige werkcollege de code om een connectie met het bestand go_sales.sqlite te maken.

In [2]:
conn = sqlite3.connect("C:\\Users\\xande\\School\\DEAI_Portfolio\\sqlitemapweek1-3\\go_sales.sqlite")
print("Database verbinding succesvol!")

Database verbinding succesvol!


Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- returned_item
- return_reason

In [3]:

#dit is alles behalve order_method.

sql_query = "SELECT * FROM product"
product = pd.read_sql_query(sql_query, conn)
sql_query = "SELECT * FROM product_type"
product_type = pd.read_sql_query(sql_query, conn)
sql_query = "SELECT * FROM product_line"
product_line = pd.read_sql_query(sql_query, conn)
sql_query = "SELECT * FROM sales_staff"
sales_staff = pd.read_sql_query(sql_query, conn)
sql_query = "SELECT * FROM sales_branch"
sales_branch = pd.read_sql_query(sql_query, conn)
sql_query = "SELECT * FROM retailer_site"
retailer_site = pd.read_sql_query(sql_query, conn)
sql_query = "SELECT * FROM country"
country = pd.read_sql_query(sql_query, conn)
sql_query = "SELECT * FROM order_header"
order_header = pd.read_sql_query(sql_query, conn)
sql_query = "SELECT * FROM order_details"
order_details = pd.read_sql_query(sql_query, conn)
sql_query = "SELECT * FROM returned_item"
returned_item = pd.read_sql_query(sql_query, conn)
sql_query = "SELECT * FROM return_reason"
return_reason = pd.read_sql_query(sql_query, conn)
print("Tabellen succesvol ingelezen!")

Tabellen succesvol ingelezen!


Krijg je een "no such table" error? Dan heb je misschien met .connect() per ongeluk een leeg  databasebestand (.sqlite) aangemaakt.<br> <u>Let op:</u> lees eventueel de informatie uit het vorige werkcollege nog eens goed door.

Als je tijdens onderstaande opdrachten uit het oog verliest welke tabellen er allemaal zijn, kan je deze Pythoncode uitvoeren:

In [4]:
sql_query = "SELECT * FROM sqlite_master WHERE type='table';"
#Vul dit codeblok verder in
pd.read_sql(sql_query, conn)
#Op de puntjes hoort de connectie naar go_sales óf go_staff óf go_crm te staan.

,type,name,tbl_name,rootpage,sql
0,table,country,country,480,"CREATE TABLE ""country"" (\n\t""COUNTRY_CODE""\tIN..."
1,table,order_details,order_details,11,"CREATE TABLE ""order_details"" (\n\t""ORDER_DETAI..."
2,table,order_header,order_header,24,"CREATE TABLE ""order_header"" (\n\t""ORDER_NUMBER..."
3,table,order_method,order_method,27,"CREATE TABLE ""order_method"" (\n\t""ORDER_METHOD..."
4,table,product,product,28,"CREATE TABLE ""product"" (\n\t""PRODUCT_NUMBER""\t..."
5,table,product_line,product_line,17,"CREATE TABLE ""product_line"" (\n\t""PRODUCT_LINE..."
6,table,product_type,product_type,10,"CREATE TABLE ""product_type"" (\n\t""PRODUCT_TYPE..."
7,table,retailer_site,retailer_site,18,"CREATE TABLE ""retailer_site"" (\n\t""RETAILER_SI..."
8,table,return_reason,return_reason,30,"CREATE TABLE ""return_reason"" (\n\t""RETURN_REAS..."
9,table,returned_item,returned_item,31,"CREATE TABLE ""returned_item"" (\n\t""RETURN_CODE..."


Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [5]:
#PRODUCT_TYPE_CODE, PRODUCTION_COST, PRODUCT_NAME.
pd.merge(product, product_type, on="PRODUCT_TYPE_CODE").loc[(product["PRODUCTION_COST"] < 100) & (product["PRODUCTION_COST"] > 50), ["PRODUCT_NAME", "PRODUCTION_COST"]]

,PRODUCT_NAME,PRODUCTION_COST
5,TrailChef Deluxe Cook Set,85.11
7,TrailChef Double Flame,75.00
16,Hibernator Lite,60.00
17,Hibernator,86.00
19,Hibernator Self - Inflating Mat,78.55
22,Hibernator Camp Cot,65.33
23,Canyon Mule Climber Backpack,62.50
45,Granite Climbing Helmet,52.86
47,Husky Harness Extreme,55.00
57,Granite Hammer,57.13


Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [6]:
#MARGIN uit product (in 0.xx vorm) en PRODUCT_NAME uit product.
pd.merge(product, product_type, on="PRODUCT_TYPE_CODE").loc[(product["MARGIN"] < 0.2) | (product["MARGIN"] > 0.6), ["PRODUCT_NAME", "MARGIN"]]

,PRODUCT_NAME,MARGIN
20,Hibernator Pad,0.17
23,Canyon Mule Climber Backpack,0.17
85,BugShield Natural,0.67
86,BugShield Spray,0.67
87,BugShield Lotion Lite,0.70
88,BugShield Lotion,0.63
89,BugShield Extreme,0.63


Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [7]:
#CURRENCY_NAME  en COUNTRY uit country.
country.loc[country["CURRENCY_NAME"] == "francs", ["COUNTRY"]].sort_values(by="COUNTRY")

,COUNTRY
15,Belgium
0,France
7,Switzerland


Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [8]:
pd.merge(product, product_type, on="PRODUCT_TYPE_CODE").loc[product["MARGIN"] > 0.5, ["INTRODUCTION_DATE"]].drop_duplicates().sort_values(by="INTRODUCTION_DATE")

,INTRODUCTION_DATE
85,1995-02-15
66,2000-10-26
101,2003-12-10
110,2003-12-15
103,2003-12-18
104,2003-12-27
106,2004-01-13


Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 7 rijen)

In [9]:
"""
Index(['SALES_BRANCH_CODE', 'ADDRESS1', 'ADDRESS2', 'CITY', 'REGION',
       'POSTAL_ZONE', 'COUNTRY_CODE'],
      dtype='str')
"""
sales_branch.loc[(sales_branch["ADDRESS2"].notna()) & (sales_branch["REGION"].notna()), ["ADDRESS1", "CITY"]].head(7)

,ADDRESS1,CITY
2,Singelgravenplein 4,Amsterdam
13,Prol. Paseo de la Reforma No. 51,Distrito Federal
14,202-2-3 Hyakunincho,Tokyo
15,543-225 Asahi,Osaka City
16,2315 Queen's Ave,Melbourne
18,"Avenida Paulista, 333",São Paulo
24,3 Albany Court,Birmingham


Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [10]:
country.loc[(country["CURRENCY_NAME"] == "dollars") | (country["CURRENCY_NAME"] == "new dollar"), ["COUNTRY"]].sort_values(by="COUNTRY")

,COUNTRY
14,Australia
3,Canada
11,Taiwan
2,United States


Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [11]:
"""
Index(['RETAILER_SITE_CODE', 'RETAILER_CODE', 'ADDRESS1', 'ADDRESS2', 'CITY',
       'REGION', 'POSTAL_ZONE', 'COUNTRY_CODE', 'ACTIVE_INDICATOR'],
      dtype='str')
"""
retailer_site.loc[(retailer_site["POSTAL_ZONE"].str.startswith("D")) & (retailer_site["ADDRESS2"].notna()), ["ADDRESS1", "ADDRESS2", "CITY"]].head(2)

,ADDRESS1,ADDRESS2,CITY
31,Röntgenstraße 90,3. Tür rechts,Frankfurt
59,Grubesallee 141,4. Stock,Hamburg


## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [12]:
"""
order_details:['ORDER_DETAIL_CODE', 'ORDER_NUMBER', 'PRODUCT_NUMBER', 'QUANTITY',
       'UNIT_COST', 'UNIT_PRICE', 'UNIT_SALE_PRICE'] 
returned_item:['RETURN_CODE', 'RETURN_DATE', 'ORDER_DETAIL_CODE', 'RETURN_REASON_CODE',
       'RETURN_QUANTITY']
"""
pd.merge(order_details, returned_item, on="ORDER_DETAIL_CODE").loc[:, ["RETURN_QUANTITY"]].sum()

RETURN_QUANTITY    11908
dtype: int64

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [13]:
sales_branch["REGION"].nunique()

15

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [14]:
pd.merge(product, product_type, on="PRODUCT_TYPE_CODE").loc[:, ["MARGIN"]].agg(["min", "max", "mean"]).round(2).T.rename(columns={"min": "Laagste marge", "max": "Hoogste marge", "mean": "Gemiddelde marge"})

,Laagste marge,Hoogste marge,Gemiddelde marge
MARGIN,0.17,0.7,0.4


Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [15]:
retailer_site.loc[retailer_site["ADDRESS2"].isna(), ["RETAILER_SITE_CODE"]].count()

RETAILER_SITE_CODE    286
dtype: int64

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [16]:
pd.merge(order_details, returned_item, on="ORDER_DETAIL_CODE").loc[order_details["UNIT_SALE_PRICE"] < order_details["UNIT_PRICE"], ["UNIT_COST"]].mean().round(2)

UNIT_COST    82.32
dtype: float64

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [17]:
"""
sales_staff: ['SALES_STAFF_CODE', 'FIRST_NAME', 'LAST_NAME', 'POSITION_EN',
       'WORK_PHONE', 'EXTENSION', 'FAX', 'EMAIL', 'DATE_HIRED',
       'SALES_BRANCH_CODE']
"""
pd.merge(sales_staff, sales_branch, on="SALES_BRANCH_CODE").loc[:, ["POSITION_EN"]].value_counts().reset_index().rename(columns={"POSITION_EN": "Aantal medewerkers", "index": "Medewerkersfunctie"})

,Aantal medewerkers,count
0,Level 2 Sales Representative,39
1,Level 3 Sales Representative,24
2,Branch Manager,19
3,Level 1 Sales Representative,13
4,District Manager,4
5,Regional Manager,2
6,General Manager,1


Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [18]:
"""
CREATE TABLE "sales_branch" (
	"SALES_BRANCH_CODE"	INTEGER,
	"ADDRESS1"	TEXT,
	"ADDRESS2"	TEXT,
	"CITY"	TEXT,
	"REGION"	TEXT,
	"POSTAL_ZONE"	TEXT,
	"COUNTRY_CODE"	INTEGER
)
CREATE TABLE "sales_staff" (
	"SALES_STAFF_CODE"	INTEGER,
	"FIRST_NAME"	TEXT,
	"LAST_NAME"	TEXT,
	"POSITION_EN"	TEXT,
	"WORK_PHONE"	TEXT,
	"EXTENSION"	INTEGER,
	"FAX"	TEXT,
	"EMAIL"	TEXT,
	"DATE_HIRED"	TEXT,
	"SALES_BRANCH_CODE"	INTEGER
)
"""
resultaat = (
    sales_staff['WORK_PHONE']
    .value_counts()
    .reset_index()
)

resultaat.columns = ['Telefoonnummer', 'Aantal medewerkers']

eind_overzicht = (
    resultaat
    .loc[resultaat['Aantal medewerkers'] > 4]
)

eind_overzicht

,Telefoonnummer,Aantal medewerkers
0,1 (617) 268-6754,6
1,1 (305) 557-4810,6
2,+33 1 68 94 52 20,5
3,+31 (0)20 692 93 94,5
4,1 (403) 232-5986,5
5,1 (416) 493-5595,5
6,1 (206) 292-0012,5
7,1 (310) 281-5722,5
8,+(61) 03 2982 4242,5
9,+(44) 121 3505267,5


## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [26]:
pd.merge(retailer_site, country, on="COUNTRY_CODE").loc[lambda x: x["COUNTRY"] == "Netherlands", ["ADDRESS1", "CITY"]].head(20)

,ADDRESS1,CITY
240,Beets 36,Beets
241,Perestraat 1,Deventer
242,Laan van Meerdervoort 966,Den Haag
243,Kalfjeslaan 47,Amstelveen
244,Startbaan 11,Amstelveen
245,Noorderspoorsingel 28,Groningen
246,Hogehilweg 13,Amsterdam
247,Spoorstraat 43,Varsseveld
248,Westhavenweg 8,Amsterdam
249,Westzeedijk 4,Muiden


Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [29]:
#Product_TYPE_EN uit product_type = Eyewear en PRODUCT_NAME uit product.
pd.merge(product, product_type, on="PRODUCT_TYPE_CODE").loc[lambda x: x["PRODUCT_TYPE_EN"] == "Eyewear", ["PRODUCT_NAME"]].head(5)

,PRODUCT_NAME
67,Polar Sun
68,Polar Ice
69,Polar Sports
70,Polar Wave
71,Polar Extreme


Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [34]:
# sales_staff heeft geen RETAILER_SITE_CODE, order_header heeft wel sales_staff_code en retailer_site_code.
pd.merge(pd.merge(order_header, sales_staff, left_on="SALES_STAFF_CODE", right_on="SALES_STAFF_CODE"), retailer_site, left_on="RETAILER_SITE_CODE", right_on="RETAILER_SITE_CODE").loc[lambda x: x["POSITION_EN"] == "Branch Manager", ["ADDRESS1", "FIRST_NAME", "LAST_NAME"]].drop_duplicates().head(1)

,ADDRESS1,FIRST_NAME,LAST_NAME
3185,"Alameda Santos, 9876",Bayard,Lopes


Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 7 rijen) 

In [49]:
#filter eerst de sales_staff op naam contains manager, daarna linken aan order_header om de verkoopdatum te krijgen. Daarna alleen de unieke posities tonen.
resultaat = (
    pd.merge(order_header, sales_staff, left_on="SALES_STAFF_CODE", right_on="SALES_STAFF_CODE")
)
eind_overzicht = (
    resultaat
    .loc[resultaat['POSITION_EN'].str.contains('manager', case=False, na=False)]
    .loc[:, ['ORDER_DATE', 'POSITION_EN']]
)
eind_overzicht

,ORDER_DATE,POSITION_EN
3185,2024-09-09,Branch Manager
3186,2024-03-13,Branch Manager
3187,2024-01-09,Branch Manager


Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [53]:
#product["PRODUCT_NAME"], product_type["PRODUCT_TYPE_EN"], Waarvan ooit meer dan 750 sutks tegelijk verkocht zijn.
#order_details["QUANTITY"] > 750. order_details["PRODUCT_NUMBER"] product["PRODUCT_NUMBER"].
resultaat = (
    pd.merge(order_details, product, left_on="PRODUCT_NUMBER", right_on="PRODUCT_NUMBER")
    .merge(product_type, left_on="PRODUCT_TYPE_CODE", right_on="PRODUCT_TYPE_CODE")
)
resultaat.loc[resultaat["QUANTITY"] > 750, ["PRODUCT_NAME", "PRODUCT_TYPE_EN"]].drop_duplicates()


,PRODUCT_NAME,PRODUCT_TYPE_EN
539,BugShield Extreme,Insect Repellents
1759,Star Peg,Tents
20121,BugShield Lotion Lite,Insect Repellents
20147,Sun Shelter 15,Sunscreen
20152,Sun Shelter 30,Sunscreen
31236,BugShield Spray,Insect Repellents
37272,Firefly 2,Lanterns
37348,BugShield Lotion,Insect Repellents
37363,Sun Shield,Sunscreen


Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [55]:
resultaat = (
    pd.merge(order_details, product, left_on="PRODUCT_NUMBER", right_on="PRODUCT_NUMBER")
)
resultaat.loc[((resultaat["UNIT_PRICE"] - resultaat["UNIT_SALE_PRICE"]) / resultaat["UNIT_PRICE"]) > 0.4, ["PRODUCT_NAME"]].drop_duplicates()

,PRODUCT_NAME
506,BugShield Natural
511,BugShield Spray
650,Hailstorm Titanium Woods Set
2797,BugShield Lotion Lite
2836,Sun Shelter 30
3751,BugShield Extreme
5074,Sun Shelter 15
6930,BugShield Lotion


Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [56]:
#return_reason["RETURN_DESCRIPTION_EN"], waarvaan ooit meer dan 90% van de afgeschafte hoeveelheid is teruggebracht.
#returned_item["RETURN_QUANTITY"] / order_details["QUANTITY"] > 0.9. return_reason["RETURN_REASON_CODE"] returned_item["RETURN_REASON_CODE"].
resultaat = (
    pd.merge(returned_item, order_details, on="ORDER_DETAIL_CODE")
    .merge(return_reason, left_on="RETURN_REASON_CODE", right_on="RETURN_REASON_CODE")
)
resultaat.loc[(resultaat["RETURN_QUANTITY"] / resultaat["QUANTITY"]) > 0.9, ["RETURN_DESCRIPTION_EN"]].drop_duplicates()

,RETURN_DESCRIPTION_EN
0,Unsatisfactory product
2,Wrong product shipped
3,Wrong product ordered


## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [58]:
#per product_type het aantal producten die daarabij behoren.
pd.merge(product, product_type, on="PRODUCT_TYPE_CODE").loc[:, ["PRODUCT_TYPE_EN"]].value_counts().reset_index().rename(columns={"PRODUCT_TYPE_EN": "Producttype", 0: "Aantal producten"})

,Producttype,count
0,Lanterns,12
1,Cooking Gear,10
2,Sleeping Bags,7
3,Climbing Accessories,7
4,Tents,6
5,Packs,6
6,Tools,6
7,Watches,5
8,Eyewear,5
9,Knives,5


Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [59]:
#per country het aantal retailers die daar gevestigd zijn.
pd.merge(retailer_site, country, on="COUNTRY_CODE").loc[:, ["COUNTRY"]].value_counts().reset_index().rename(columns={"COUNTRY": "Land", 0: "Aantal retailers"})

,Land,count
0,United States,101
1,Germany,32
2,Canada,31
3,France,29
4,Japan,25
5,United Kingdom,25
6,Netherlands,20
7,Switzerland,13
8,Sweden,12
9,Italy,12


Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (6 kolommen, 10 rijen)

In [62]:
#product_type["PRODUCT_TYPE_EN"] = "Cookin Gear"
#per product["PRODUCT_NAME"] sum verkochte hoeveelheid en gemiddelde verkoopprijs.
resultaat = (
    pd.merge(product, product_type, on="PRODUCT_TYPE_CODE")
    .merge(order_details, left_on="PRODUCT_NUMBER", right_on="PRODUCT_NUMBER")
)
resultaat = resultaat[resultaat["PRODUCT_TYPE_EN"] == "Cooking Gear"]
resultaat.groupby("PRODUCT_NAME").agg({"QUANTITY": "sum", "UNIT_SALE_PRICE": "mean"}).reset_index().rename(columns={"QUANTITY": "Totale verkochte hoeveelheid", "UNIT_SALE_PRICE": "Gemiddelde verkoopprijs"}).sort_values(by="Totale verkochte hoeveelheid", ascending=False)

,PRODUCT_NAME,Totale verkochte hoeveelheid,Gemiddelde verkoopprijs
9,TrailChef Water Bag,36738,6.032506
2,TrailChef Cup,27418,6.976490
0,TrailChef Canteen,26544,11.979418
6,TrailChef Kitchen Kit,22206,22.345975
8,TrailChef Utensils,13776,17.294151
4,TrailChef Double Flame,12476,132.672882
1,TrailChef Cook Set,10878,51.687926
5,TrailChef Kettle,8826,12.201235
7,TrailChef Single Flame,8504,63.827622
3,TrailChef Deluxe Cook Set,3110,121.217638


Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [76]:
#per country country["COUNTRY"], retailer_site["CITY"] waar de verkoopafdeling is gevestigt als verkoper op COUNTRY_CODE.
#verder, aantal CITY waar zich klanten bevinden (kolomnaam 'klanten')
resultaat = pd.merge(retailer_site, country, on="COUNTRY_CODE")
resultaat.groupby(["COUNTRY", "CITY"]).agg({"RETAILER_CODE": "nunique"}).reset_index().rename(columns={"COUNTRY": "Land", "CITY": "verkopers", "RETAILER_CODE": "klanten"}).sort_values(by="klanten", ascending=False)

,Land,verkopers,klanten
29,Canada,Montréal,6
77,Germany,Frankfurt,6
66,France,Paris,5
102,Japan,Tokyo,5
166,United Kingdom,London,5
...,...,...,...
220,United States,Philadelphia,1
230,United States,Springfield,1
231,United States,Topeka,1
232,United States,Tucson,1


## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [78]:
medewerkers_zonder_verkoop = pd.merge(sales_staff, order_header, left_on="SALES_STAFF_CODE", right_on="SALES_STAFF_CODE", how="left", indicator=True)
medewerkers_zonder_verkoop = medewerkers_zonder_verkoop[medewerkers_zonder_verkoop["_merge"] == "left_only"]
resultaat = medewerkers_zonder_verkoop[["FIRST_NAME", "LAST_NAME"]]
resultaat

,FIRST_NAME,LAST_NAME
0,Denis,Pagé
320,Else,Mörike
321,Frank,Fuchs
647,Fritz,Hirsch
648,Jörg,Kunze
728,Maria,Iacobucci
946,Kick,Kalkman
1309,Karin,Bergström
1395,Sally,White
1421,Frank,Bretton


Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [ ]:
gemiddelde_marge = product["MARGIN"].mean()
producten_lagere_marge = product[product["MARGIN"] < gemiddelde_marge]
aantal_producten = len(producten_lagere_marge)
resultaat = pd.DataFrame({"Aantal producten met lagere marge": [aantal_producten], "Gemiddelde marge van deze producten": [producten_lagere_marge["MARGIN"].mean()]})
resultaat

,Aantal producten met lagere marge,Gemiddelde marge van deze producten
0,59,0.292203


Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [81]:
resultaat = (
    pd.merge(order_details, returned_item, on="ORDER_DETAIL_CODE", how="left", indicator=True)
    .merge(product, left_on="PRODUCT_NUMBER", right_on="PRODUCT_NUMBER")
)
resultaat = resultaat[(resultaat["UNIT_SALE_PRICE"] > 500) & (resultaat["QUANTITY"] > 0) & (resultaat["_merge"] == "left_only")]
resultaat.loc[:, ["PRODUCT_NAME"]].drop_duplicates() 

,PRODUCT_NAME
88,Star Dome
95,Star Gazer 2
103,Star Gazer 3
112,Star Gazer 6
616,Hailstorm Steel Irons
625,Hailstorm Titanium Irons
634,Lady Hailstorm Steel Irons
641,Lady Hailstorm Titanium Irons
647,Hailstorm Titanium Woods Set
657,Hailstorm Steel Woods Set


Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [87]:
#per LAST_NAME of POSITION_EN manager bevat in een ja/nee kolom of deze medewerker een manager is.
#for loop sales_staff, resultaat alleen 1 kolom.
resultaat = pd.DataFrame({
    "LAST_NAME": sales_staff["LAST_NAME"],
    "Is_manager": sales_staff["POSITION_EN"].apply(lambda x: "Ja" if "manager" in x.lower() else "Nee")
})
resultaat

,LAST_NAME,Is_manager
0,Pagé,Ja
1,Michel,Nee
2,Clermont,Nee
3,Jauvin,Nee
4,Wiesinger,Nee
...,...,...
97,Laermans,Nee
98,De Crée,Nee
99,Lattrez,Nee
100,Seefelder,Nee


Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [19]:
from datetime import date
date.today().year

2026

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [20]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 30 jaar in dienst) of een ‘lang in dienst’ (groter dan gelijk aan 30 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [102]:

sales_staff["DATE_HIRED"] = pd.to_datetime(sales_staff["DATE_HIRED"], format='%Y-%m-%d', errors='coerce')
huidig_jaar = date.today().year
sales_staff["jaar_in_dienst"] = (huidig_jaar - sales_staff["DATE_HIRED"].dt.year).fillna(0).astype(int)
sales_staff["Dienstverband"] = sales_staff["jaar_in_dienst"].apply(lambda x: "Kort in dienst" if x < 30 else "Lang in dienst")
resultaat = sales_staff[["jaar_in_dienst", "Dienstverband"]].sort_values(by="jaar_in_dienst", ascending=False)

resultaat

,jaar_in_dienst,Dienstverband
30,36,Lang in dienst
28,35,Lang in dienst
24,34,Lang in dienst
25,34,Lang in dienst
33,34,Lang in dienst
...,...,...
10,26,Kort in dienst
67,26,Kort in dienst
93,26,Kort in dienst
95,26,Kort in dienst
